pal/AzureIoTSocket_Ethernet/socketio_esp32ethernet2.cpp

```diff
@@ -0,0 +1,577 @@
+        if ((socket_io_instance->io_state != IO_STATE_CLOSED) && (socket_io_instance->io_state != IO_STATE_CLOSING))
+        {
+            // Only close if the socket isn't already in the closed or closing state
+            if (socket_io_instance->socket != NULL)
+            {
+                socket_io_instance->socket->stop();
+            }
+
+            if (socket_io_instance->hostname != NULL)
+            {
+                // We created this socket so delete it
+                delete socket_io_instance->socket;
+            }
+            socket_io_instance->socket = NULL;
+            socket_io_instance->io_state = IO_STATE_CLOSED;
+        }
+
+        if (on_io_close_complete != NULL)
+        {
+            on_io_close_complete(callback_context);
+        }
+        result = 0;
+    }
+
+    return result;
+}
+
+int socketio_send(CONCRETE_IO_HANDLE socket_io, const void* buffer, size_t size, ON_SEND_COMPLETE on_send_complete, void* callback_context)
+{
+    int result;
+
+    if ((socket_io == NULL) ||
+        (buffer == NULL) ||
+        (size == 0))
+    {
+        /* Invalid arguments */
+        LogError("Invalid argument: send given invalid parameter");
+        result = MU_FAILURE;
+    }
+    else
+    {
+        SOCKET_IO_INSTANCE* socket_io_instance = (SOCKET_IO_INSTANCE*)socket_io;
+        if (socket_io_instance->io_state != IO_STATE_OPEN)
+        {
+            LogError("Failure: socket state is not opened.");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            LIST_ITEM_HANDLE first_pending_io = singlylinkedlist_get_head_item(socket_io_instance->pending_io_list);
+            if (first_pending_io != NULL)
+            {
+                if (add_pending_io(socket_io_instance, (const unsigned char*)buffer, size, on_send_complete, callback_context) != 0)
+                {
+                    LogError("Failure: add_pending_io failed.");
+                    result = MU_FAILURE;
+                }
+                else
+                {
+                    result = 0;
+                }
+            }
+            else
+            {
+                ssize_t send_result = socket_io_instance->socket->write((const uint8_t *)buffer, size);
+
+                if (send_result == 0)
+                {
+                    LogError("Failure: sending socket failed. errno=%d (%s).", errno, strerror(errno));
+                    result = MU_FAILURE;
+                }
+                else if (send_result < size)
+                {
+                    /* queue data */
+                    if (add_pending_io(socket_io_instance, (const unsigned char*)((char *)buffer + send_result), size - send_result, on_send_complete, callback_context) != 0)
+                    {
+                        LogError("Failure: add_pending_io failed.");
+                        result = MU_FAILURE;
+                    }
+                    else
+                    {
+                        result = send_result;
```

Comment:
>send_result [](start = 33, length = 11)

what does the result mean?
when full buffer is send in sync report zero if half send how much is send sync?